### ISSUE TITLE
Haproxy starts more than 1 hour from server-state-file

### ISSUE THREAD

**Issue by nadezhdabelan** [haproxy/haproxy#55]
Hi
## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 1.8.3-205f675 2017/12/30
Copyright 2000-2017 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -m64 -march=x86-64 -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label
  OPTIONS = USE_SLZ=1 USE_REGPARM=1 USE_OPENSSL=1 USE_SYSTEMD=1 USE_PCRE=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
Running on OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Encrypted password support via crypt(3): yes
Built with multi-threading support.
Built with PCRE version : 8.32 2012-11-30
Running on PCRE version : 8.32 2012-11-30
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
        [SPOE] spoe
        [COMP] compression
        [TRACE] trace
Linux test 3.10.0-514.21.1.el7.x86_64 #1 SMP Thu May 25 17:04:51 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

```

## What's the configuration?

```
global
    log         127.0.0.1 local2

    chroot      /var/lib/haproxy
    pidfile     /var/run/haproxy.pid
    maxconn     4000
    user        haproxy
    group       haproxy
    daemon

    stats socket /var/run/haproxy.sock mode 600 level admin
    stats timeout 2m    
    server-state-file /tmp/server_state

    ssl-default-bind-ciphers PROFILE=SYSTEM
    ssl-default-server-ciphers PROFILE=SYSTEM


defaults
    mode                    http
    log                     global
    option                  httplog
    option                  dontlognull
    option http-server-close
    option forwardfor       except 127.0.0.0/8
    option                  redispatch
    retries                 3
    timeout http-request    10s
    timeout queue           1m
    timeout connect         10s
    timeout client          1m
    timeout server          1m
    timeout http-keep-alive 10s
    timeout check           10s
    maxconn                 3000
    load-server-state-from-file global

#---------------------------------------------------------------------
# main frontend which proxys to the backends
#---------------------------------------------------------------------
frontend main
    bind *:5000
    acl url_static       path_beg       -i /static /images /javascript /stylesheets
    acl url_static       path_end       -i .jpg .gif .png .css .js

    use_backend static          if url_static
    default_backend             app

frontend https_frontend
        bind *:443
        mode    tcp
        timeout client 50000ms
        tcp-request inspect-delay 10s
        # Accept only TLS connections
        tcp-request content accept   if { req_ssl_hello_type 1 }
        tcp-request content reject

        #option dontlog-normal

        # Internal networks
        acl wh_list   src -f /etc/haproxy/trusted_networks.lst
        acl int_hosts req.ssl_sni,map(/etc/haproxy/hosts_https_int.map) -m found
        

        # public resources
        use_backend %[req.ssl_sni(),lower,map(/etc/haproxy/hosts_https.map)]

listen stats # Define a listen section called "stats"
    bind :9000 # Listen on localhost:9000
    mode http
    stats enable  # Enable stats page
    stats hide-version  # Hide HAProxy version
    stats realm Haproxy\ Statistics  # Title text for popup window
    stats uri /haproxy_stats  # Stats URI
    #stats auth Username:Password  # Authentication credentials


#---------------------------------------------------------------------
# static backend for serving up images, stylesheets and such
#---------------------------------------------------------------------
backend static
    balance     roundrobin
    server      static 127.0.0.1:4331 check

backend link1
	mode tcp
	server tunnelsrv 127.0.0.1:5000 
backend link2
	mode tcp
	server tunnelsrv 127.0.0.1:5000 
backend link3
	mode tcp
	server tunnelsrv 127.0.0.1:5000 
backend link4
	mode tcp
	server tunnelsrv 127.0.0.1:5000 
backend link5
	mode tcp
	server tunnelsrv 127.0.0.1:5000
......
backend link99999
	mode tcp
	server tunnelsrv 127.0.0.1:5000 
```

## Steps to reproduce the behavior

1.  socat /var/run/haproxy.sock - <<< "show servers state" > /tmp/server_state
2. systemctl start haproxy.service

## Actual behavior

Error is that service doesnot start during more than 1 hour
Without config ```:server-state-file /tmp/server_state``` it starts during 5 sec

## Expected behavior
Service starts during 5 sec

## Do you have any idea what may have caused this?
Possible it ping all 99999 backends
## Do you have an idea how to solve the issue?

no

**Comment by wtarreau** [comment]
There's an O(N^2) search in the state file processing, which takes a while with many backends. I remember that @bedis has been working on it with pretty good results. CCing him just in case he can bring in any new info on this subject.

**Comment by nadezhdabelan** [comment]
> There's an O(N^2) search in the state file processing, which takes a while with many backends. I remember that @bedis has been working on it with pretty good results. CCing him just in case he can bring in any new info on this subject.

Thanks, I will write him!

**Comment by bedis** [comment]
Hi @nadezhdabelan 

I confirm the behavior is by design, as explained by @wtarreau .
I have a wok-in-progress patch which improves this greatly, but it's just a PoC and I can't share it yet.
I am finishing an other feature (do-resolve) and I'll focus on fixing server-state parsing.
I will send you my patches once they are clean.

That said, can you explain me your use-case? maybe we can improve the configuration :)

**Comment by nadezhdabelan** [comment]
> I confirm the behavior is by design, as explained by @wtarreau .
> I have a wok-in-progress patch which improves this greatly, but it's just a PoC and I can't share it yet.
> I am finishing an other feature (do-resolve) and I'll focus on fixing server-state parsing.
> I will send you my patches once they are clean.
> 
> That said, can you explain me your use-case? maybe we can improve the configuration :)

Hi, @bedis 
Thanks very much!
Yes, I confirm - backends never change their name (but backends more than 40 thousands ), they change only IP addresses through unix sockets at runtime
I know how realize our case without server-state (but it is necessary additional actions)....

So, I will wait your changes!

**Comment by wtarreau** [comment]
It was addressed as commit da29fe2 in 2.1-dev1. Now closing as fixed (and will not be backported since it's a non-trivial optimization).
